receiver/nvmlreceiver/client.go

```diff
@@ -0,0 +1,276 @@
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"go.uber.org/zap"
+)
+
+const maxWarningsForFailedDeviceMetricQuery = 5
+
+type nvmlClient struct {
+	logger                         *zap.SugaredLogger
+	disable                        bool
+	handleCleanup                  func()
+	devices                        []nvml.Device
+	devicesModelName               []string
+	deviceToLastSeenTimestamp      map[nvml.Device]uint64
+	deviceMetricToFailedQueryCount map[string]uint64
+}
+
+type nvmlMetric struct {
+	time  time.Time
+	gpuID uint
+	name  string
+	value [8]byte
+}
+
+// calling nvml.Init() twice causes an unnecessary error (also wrap here for mocking)
+var once sync.Once
+var nvmlInitReturn nvml.Return
+var nvmlInit = func() nvml.Return {
+	once.Do(func() {
+		nvmlInitReturn = nvml.Init()
+	})
+	return nvmlInitReturn
+}
+
+var nvmlDeviceGetSamples = nvml.DeviceGetSamples
+var nvmlDeviceGetMemoryInfo = nvml.DeviceGetMemoryInfo
+
+func newClient(config *Config, logger *zap.Logger) (*nvmlClient, error) {
+	nvmlCleanup, err := initializeNvml(logger)
+	if err != nil {
+		logger.Sugar().Warnf("Unable to find and/or initialize Nvidia Management Library on '%v'. No Nvidia device metrics will be collected.", err)
+		return &nvmlClient{logger: logger.Sugar(), disable: true}, nil
+	}
+
+	devices, names, err := discoverDevices(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &nvmlClient{
+		logger:                         logger.Sugar(),
+		disable:                        false,
+		handleCleanup:                  nvmlCleanup,
+		devices:                        devices,
+		devicesModelName:               names,
+		deviceToLastSeenTimestamp:      make(map[nvml.Device]uint64),
+		deviceMetricToFailedQueryCount: make(map[string]uint64),
+	}, nil
+}
+
+func initializeNvml(logger *zap.Logger) (nvmlCleanup func(), err error) {
+	nvmlCleanup = nil
+
+	defer func() {
+		// applicable to tagged releases of github.com/NVIDIA/go-nvml <= v0.11.6-0
+		if perr := recover(); perr != nil {
+			err = fmt.Errorf("%v", perr)
+		}
+	}()
+
+	ret := nvmlInit()
+	if ret != nvml.SUCCESS {
+		if ret == nvml.ERROR_LIBRARY_NOT_FOUND {
+			err = fmt.Errorf("libnvidia-ml.so not found")
+		} else {
+			err = fmt.Errorf("'%v'", nvml.ErrorString(ret))
+		}
+		return
+	}
+	logger.Sugar().Infof("Successfully initialized Nvidia Management Library")
+
+	nvmlCleanup = func() {
+		ret := nvml.Shutdown()
+		if ret != nvml.SUCCESS {
+			logger.Sugar().Infof("Unable to shutdown Nvidia Management library on '%v'", nvml.ErrorString(ret))
+		}
+	}
+
+	err = nil
+	return
+}
+
+func discoverDevices(logger *zap.Logger) ([]nvml.Device, []string, error) {
+	count, ret := nvml.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		return nil, nil, fmt.Errorf("Unable to get Nvidia device count on '%v'", nvml.ErrorString(ret))
+	}
+
+	devices := make([]nvml.Device, count)
+	names := make([]string, count)
+	for i := 0; i < count; i++ {
+		devices[i], ret = nvml.DeviceGetHandleByIndex(i)
```

Comment:
Hey, thanks for taking note of this! Yes, I agree we should add UUID as a metric attribute so that it's possible to track the same GPU across reboots.

Reply:
I added the UUID. Though on second thought I'm not sure if there is much value to tracking the UUID between reboots. In any case, adding the UUID doesn't affect the cardinality. So why not. I don't think we want to sort on UUID as it would be confusing to have gpu_number not match nvidia-smi.

Reply:
This looks good to me. Thanks for making the change.